lib/metasploit/framework/login_scanner/jenkins.rb

```diff
@@ -5,50 +5,133 @@ module Framework
+        end
+
+        private
+
+        # This method takes a username and password and a target URI
+        # then attempts to login to Jenkins and will either fail with appropriate errors
+        #
+        # @param [String] username The username for login credentials
+        # @param [String] password The password for login credentials
+        # @return [Array] [status, proof] The result of the login attempt
+        def jenkins_login(username, password)
+          begin
+            res = send_request(
               'method' => method,
               'uri' => uri,
-              'vars_post' => request['vars_post']
-            })
+              'vars_post' => {
+                'j_username' => username,
+                'j_password' => password,
+                'Submit' => 'log in'
+              }
+            )
+
+            if res && res.headers['location'] && !res.headers['location'].include?('loginError')
+              status = Metasploit::Model::Login::Status::SUCCESSFUL
+              proof = res.headers
+            else
+              status = Metasploit::Model::Login::Status::INCORRECT
+              proof = res
+            end
+          rescue ::EOFError, Errno::ETIMEDOUT, Errno::ECONNRESET, Rex::ConnectionError, OpenSSL::SSL::SSLError, ::Timeout::Error => e
+            status = Metasploit::Model::Login::Status::UNABLE_TO_CONNECT
+            proof = e
           end
 
-          result_opts.merge!(status: status, proof: proof)
+          [status, proof]
+        end
 
-          Result.new(result_opts)
+        # This method uses the provided URI to determine whether login is possible for Jenkins.
+        # Based on the contents of the provided URI, the method looks for the login form and
+        # extracts the endpoint used to authenticate against.
+        #
+        # @param [URI, String] target_uri The targets URI
+        # @return [String, nil] URI for successful login
+        def jenkins_uri_check(target_uri, keep_cookies: false)
+          # if keep_cookies is true we get the first cookie that's needed by newer Jenkins versions
+          res = send_request(
+            'method' => 'GET',
+            'uri' => normalize_uri(target_uri, 'login'),
+            'keep_cookies' => keep_cookies
+          )
+
+          return normalize_uri(target_uri, JENKINS_LOGIN_VALIDATION_URL) unless valid_response?(res)
+
+          if res&.body =~ /action="(j_([a-z0-9_]+))"/
+            uri = Regexp.last_match(1)
+            normalize_uri(target_uri, uri)
+          else
+            normalize_uri(target_uri, JENKINS_LOGIN_VALIDATION_URL)
+          end
+        end
+
+        # Determines whether the provided response is considered valid or not.
+        #
+        # @param [Rex::Proto::Http::Response, nil] response The response received from the HTTP request.
+        # @return [Boolean] True if the response if valid; otherwise false.
+        def valid_response?(response)
+          response&.code == 200
```

Comment:
should this compare to `self.http_success_codes` instead of hardcoding 200?

Reply:
For Jenkins specific behavior I don't see the point of making this even configurable.

Reply:
just thinking for consistency, someone later down the line might expect it to be making use of `http_success_codes` like the rest of the http login scanners, I don't like the idea of making some scanners ignoring it and doing their own thing just because we know that at the moment only a 200 is valid